### PR TITLE
Add FieldBuilder interface and related implementations 

### PR DESCRIPTION
**Pull request by guawang**
1. Add FieldBuilder interface and related implementations in impls 1,4 - 1.9.
2. Add FIELD_BUILDER instance and  one method in AvroCompatibilityHelper to create a Schema.Field by cloning from existing Field.

### CODE REVIEW

File: helper/helper/src/main/java/com/linkedin/avroutil1/compatibility/AvroCompatibilityHelper.java [hunk 1]
```diff
@@ -246,12 +260,7 @@ public static JsonDecoder newJsonDecoder(Schema schema, String in) throws IOExce
 
   /**
    * constructs a {@link JsonDecoder} on top of the given {@link InputStream} for the given {@link Schema}
-   * that is more widely compatible than the "native" avro decoder:
```
Comment (radai-rosenblatt) [inline helper/helper/src/main/java/com/linkedin/avroutil1/compatibility/AvroCompatibilityHelper.java:249]: why strip these docs out ?

File: helper/helper/src/main/java/com/linkedin/avroutil1/compatibility/AvroCompatibilityHelper.java [hunk 2]
```diff
@@ -27,20 +39,14 @@
...
-import com.linkedin.avroutil1.compatibility.avro19.Avro19Adapter;
-
 
 /**
  * a Utility class for performing various avro-related operations under a wide range of avro versions at runtime.
  */
 public class AvroCompatibilityHelper {
   private static final AvroVersion DETECTED_VERSION;
   private static final AvroAdapter ADAPTER;
+  private static final FieldBuilder FIELD_BUILDER;
```
Comment (radai-rosenblatt) [inline helper/helper/src/main/java/com/linkedin/avroutil1/compatibility/AvroCompatibilityHelper.java:49]: why is this required? why not add a method - cloneField() into the adapter interface that returns a field builder?

File: helper/helper/src/main/java/com/linkedin/avroutil1/compatibility/AvroCompatibilityHelper.java [hunk 3]
```diff
@@ -453,6 +462,18 @@ public static Object getGenericDefaultValue(Schema.Field field) {
...
   }
 
+  /**
+   * returns a new schema field, cloned from an existing schema field and a given schema.
+   * @param field a schema field
+   * @param schema a schema
+   * @return a new field that cloned from the given field and schema
+   * @throws org.apache.avro.AvroRuntimeException if the field in question has no default.
+   */
+  public static Schema.Field cloneSchemaField(Schema.Field field, Schema schema) {
```
Comment (radai-rosenblatt) [inline helper/helper/src/main/java/com/linkedin/avroutil1/compatibility/AvroCompatibilityHelper.java:472]: i think a cleaner API would be:
`public static FieldBuilder cloneField(Schema.Field original)`
users get a builder, and can override any properties they like and call build()

File: helper/impls/helper-impl-17/src/main/java/com/linkedin/avroutil1/compatibility/avro17/FieldBuilder17.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+package com.linkedin.avroutil1.compatibility.avro17;
+
+import com.linkedin.avroutil1.compatibility.FieldBuilder;
+import org.apache.avro.Schema;
+import org.apache.avro.Schema.Field.Order;
+import org.codehaus.jackson.JsonNode;
+
+
+public class FieldBuilder17 implements FieldBuilder {
+  private static Schema.Field _field;
```
Comment (radai-rosenblatt) [inline helper/impls/helper-impl-17/src/main/java/com/linkedin/avroutil1/compatibility/avro17/FieldBuilder17.java:16]: why is this field static? also i think the Schema.Field (original) should be final and set at constructor time for the builder

Comment (guawang) [reply helper/impls/helper-impl-17/src/main/java/com/linkedin/avroutil1/compatibility/avro17/FieldBuilder17.java:16]: yes it should be final but not static.

File: helper/helper/src/main/java/com/linkedin/avroutil1/compatibility/AvroCompatibilityHelper.java (hunk 3 above)
Comment (guawang) [reply helper/helper/src/main/java/com/linkedin/avroutil1/compatibility/AvroCompatibilityHelper.java:472]: I thought for avro-utils, we should not return the "internal" class but return a type in AVRO libraries, like Schema, or Decoder etc.

File: helper/helper/src/main/java/com/linkedin/avroutil1/compatibility/AvroCompatibilityHelper.java (hunk 2 above)
Comment (guawang) [reply helper/helper/src/main/java/com/linkedin/avroutil1/compatibility/AvroCompatibilityHelper.java:49]: I thought for avro-utils, we should not return the "internal" class but return a type in AVRO libraries, like Schema, or Decoder etc.

File: helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/FieldBuilder.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+package com.linkedin.avroutil1.compatibility;
+
+import org.apache.avro.Schema;
+import org.apache.avro.Schema.Field.Order;
+
+
+/**
+ * an builder interface to build a Schema.Field with given parameters.
+ * The defaultValue needed to be defined in implemented classes since default
+ * value type is changed to Object since AVRO 1.9
```
Comment (radai-rosenblatt) [inline helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/FieldBuilder.java:16]: we could support getting/setting the default value as either Object (primitive-wrapper collection or GenericData.Record), and/or as a json string (cat least as a todo)

**Review by radai-rosenblatt — APPROVED** [review]
